substrate-node/pallets/pallet-tfgrid/src/migrations/v14.rs

```diff
@@ -0,0 +1,203 @@
+        );
+
+        info!("👥  TFGrid pallet to V14 passes PRE migrate checks ✅",);
+        Ok(())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        if PalletVersion::<T>::get() == StorageVersion::V13Struct {
+            migrate_nodes::<T>() + migrate_farms::<T>()
+        } else {
+            info!(" >>> Unused migration");
+            0
+        }
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade() -> Result<(), &'static str> {
+        assert!(PalletVersion::<T>::get() >= types::StorageVersion::V14);
+        // Check number of nodes against pre-check result
+        let pre_nodes_count = Self::get_temp_storage("pre_node_count").unwrap_or(0u64);
+        assert_eq!(
+            Nodes::<T>::iter().count() as u64,
+            pre_nodes_count,
+            "Number of nodes migrated does not match"
+        );
+
+        info!(
+            "👥  FixPublicIPV13 post migration: migration to {:?} passes POST migrate checks ✅",
+            PalletVersion::<T>::get()
+        );
+
+        Ok(())
+    }
+}
+
+pub fn migrate_nodes<T: Config>() -> frame_support::weights::Weight {
+    info!(
+        " >>> Starting tfgrid pallet migration, pallet version: {:?}",
+        PalletVersion::<T>::get()
+    );
+
+    let mut migrated_count = 0;
+
+    Nodes::<T>::translate::<
+        super::types::v13::Node<LocationOf<T>, InterfaceOf<T>, SerialNumberOf<T>>,
+        _,
+    >(|k, n| {
+        // By default initialize the public config to None
+        let mut public_config = None;
+        if let Some(config) = n.public_config {
+            // If the config is valid, keep it, otherwise discard
+            match config.is_valid() {
+                Ok(_) => public_config = Some(config),
+                Err(_) => {
+                    debug!("resetting pub config of node: {:?}", k);
+                    public_config = None;
+                }
+            }
+        }
+
+        let new_node = Node {
+            version: TFGRID_NODE_VERSION,
+            id: n.id,
+            farm_id: n.farm_id,
+            twin_id: n.twin_id,
+            resources: n.resources,
+            location: n.location,
+            power: n.power,
+            public_config,
+            created: n.created,
+            farming_policy_id: n.farming_policy_id,
+            interfaces: n.interfaces,
+            certification: n.certification,
+            secure_boot: n.secure_boot,
+            virtualized: n.virtualized,
+            serial_number: n.serial_number,
+            connection_price: n.connection_price,
+        };
+        debug!("Node: {:?} succesfully migrated", k);
+        migrated_count += 1;
+        Some(new_node)
+    });
+
+    info!(
+        " <<< Node storage updated! Migrated {} Nodes ✅",
+        migrated_count
+    );
+
+    // Return the weight consumed by the migration.
+    T::DbWeight::get().reads_writes(migrated_count as Weight + 1, migrated_count as Weight + 1)
+}
+
+use sp_std::vec;
```

Comment:
use statements outside of functions or other blocks are by convention grouped at the top of the file